### PR TITLE
Enable C++11 and 14

### PR DESCRIPTION
**Pull request by seantalts**
Enable C++11 and available C++14 extensions in the build systems. 
#### Copyright and Licensing

Please list the copyright holder for the work you are submitting (this will be you or your assignee, such as a university or company): Columbia University

By submitting this pull request, the copyright holder is agreeing to license the submitted work under the following licenses:
- Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)
- Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)

### CODE REVIEW

**Comment by seantalts** [comment]
I think this is ready for review. @syclik would you mind taking a look?

Some context: It's enabling c++14 and not checking that we only use code that aligns with the select compilers we use. However, it is testing with gcc 4.9, which is the limiting factor on most (all?) of the c++14 adoption, so I think this is fine.

File: make/os_mac [hunk 1]
```diff
@@ -14,17 +14,15 @@ ifeq (g++,$(CC_TYPE))
...
-  CFLAGS += -Wno-c++11-long-long
   CFLAGS += -Wsign-compare
   CFLAGS += -Wno-unused-local-typedef
-  ifeq (true,$(C++11))
-    CFLAGS += -stdlib=libc++ -std=c++11
-  endif
+
+  #https://stackoverflow.com/questions/19774778/when-is-it-necessary-to-use-use-the-flag-stdlib-libstdc
+  #for OSX before mavericks, specifically, but fine for post-mavericks too
+  #CXXFLAGS += -stdlib=libc++
```
Comment (syclik) [inline make/os_mac]: Remove these lines? It's already added in `make/detect_cc`

File: make/os_mac [hunk 2]
```diff
@@ -14,17 +14,15 @@ ifeq (g++,$(CC_TYPE))
...
   #PCH = .pch
   ifeq (4.2,$(CC_MAJOR).$(CC_MINOR))
```
Comment (syclik) [inline make/os_mac]: Let's remove this `ifeq` block. It's no longer relevant.

File: makefile [hunk 3]
```diff
@@ -31,7 +30,8 @@ include make/libraries
 # Set default compiler options.
 ##
 CFLAGS = -I . -isystem $(EIGEN) -isystem $(BOOST) -isystem$(CVODES)/include -Wall -DBOOST_RESULT_OF_USE_TR1 -DBOOST_NO_DECLTYPE -DBOOST_DISABLE_ASSERTS -DNO_FPRINTF_OUTPUT -pipe
-CFLAGS_GTEST = -DGTEST_USE_OWN_TR1_TUPLE
+CXXFLAGS = -std=c++14
```
Comment (syclik) [inline makefile]: why was this added to `CXX` flags instead of just `CFLAGS`?

Comment (seantalts) [inline makefile]: `CXXFLAGS` is _just_ for the C++ compiler while `CFLAGS` is just for the C compiler. This only matters because cvodes (and maybe other libraries?) are not C++. Cvodes is actually Objective C, apparently. And `CFLAGS` are sent to its compiler (creating errors) but `CXXFLAGS` are not.

File: make/os_mac (hunk 2 above)
Comment (seantalts) [inline make/os_mac]: Do we want to keep a template depth argument or just remove the entire block?

Comment (bob-carpenter) [inline make/os_mac]: The depth specification was a workaround for one of the Xcode clang++ releases that inadvertently set the depth too low.  Don't know if we still need it---it was for a pretty old release at this point.

Comment (syclik) [inline make/os_mac]: I think we should just remove it. I just took a look at that file and there's a repeated block in the g++ section that could be removed too.

File: makefile (hunk 3 above)
Comment (syclik) [inline makefile]: We're actually using `CFLAGS` and `CC` to build C++ code (as I mentioned before, it's a misuse of make, but we've been doing it since the start and there wasn't a great time to shift it to the proper use of make).

It looks like `CXXFLAGS` is only used for testing. (line 34 of makefile) Any reason not to just add `-std=c++14` directly to `CFLAGS_GTEST`? Just wanting to figure out the purpose of the indirection.

Comment (seantalts) [inline makefile]: I think it's a step in the right direction towards better use of `make` (though maybe we want to get to `cmake` or something...?). `CXXFLAGS` will be used in implicit rules and the `COMPILE.cc` macro: http://docs.oracle.com/cd/E19504-01/802-5880/6i9k05dhg/index.html

Though you are correct that I don't think we make use of that indirection right now. I'm like halfway through making the make system in the math lib distinguish between C and C++ targets more properly, but I think it'd be best to do that in a separate pull request. I might then try to include more of the math library's makefiles in the Stan makefile to reduce duplication.

**Comment by seantalts** [comment]
@syclik I think I addressed the comments if you wouldn't mind taking another look. Thanks!

**Comment by syclik** [comment]
@seantalts, thanks. I'm an idiot: I had a pending review left on github that I'll submit now (once I find the submit button). 

Thanks for addressing everything. I'll fix the remaining items that I spotted (they are minor).

**Comment by syclik** [comment]
@seantalts, I made the last couple fixes.

Please merge after tests pass. (If I see it, I'll merge.)

**Comment by seantalts** [comment]
Now cvodes is compiling with `-stdlib=libc++` on clang, which seems to work but is not ideal. I'm not sure what about using `CXXFLAGS` with `CFLAGS` is inconsistent; we'll need both even when we straighten out these makefiles. We'll also likely use a third one called `CPPFLAGS` that is sent to both C and C++ compilers. Anyway, I'd love to see this merged in. For Stan and CmdStan, would you imagine we just take these same changes and apply them upstream since they don't share the relevant parts of the makefile configuration?

**Comment by syclik** [comment]
Ah. I see. I'll fix that.

It's inconsistent because we're still compiling everything with CC instead
of CXX. I just created an issue to address it.

For Stan and CmdStan, they use includes to pull in the math makefiles, so
for the most part, we shouldn't need to duplicate.

**Comment by seantalts** [comment]
I took a look and I'm not sure these would get pulled in; the stan makefile just includes $(MATH)make/libraries from what I can tell. Also if the proper flags were only present for GTEST (as in the current iteration) they wouldn't go anywhere used to build stan programs, right?

I guess I'll go ahead and merge this and start on the cleanup PR, which I hope will put us in a better place to just include the c++11 stuff in Stan and CmdStan.
